server/container_restore.go

```diff
@@ -0,0 +1,279 @@
+	stopMutex.RLock()
+	defer stopMutex.RUnlock()
+	if sb.Stopped() {
+		return "", fmt.Errorf("CreateContainer failed as the sandbox was stopped: %s", sb.ID())
+	}
+
+	ctr, err := container.New()
+	if err != nil {
+		return "", fmt.Errorf("failed to create container: %w", err)
+	}
+
+	containerConfig := &types.ContainerConfig{
+		Metadata: &types.ContainerMetadata{
+			Name:    ctrMetadata.Name,
+			Attempt: ctrMetadata.Attempt,
+		},
+		Image: &types.ImageSpec{Image: config.RootfsImageName},
+		Linux: &types.LinuxContainerConfig{
+			Resources:       &types.LinuxContainerResources{},
+			SecurityContext: &types.LinuxContainerSecurityContext{},
+		},
+		Annotations: originalAnnotations,
+		Labels:      originalLabels,
+	}
+
+	ignoreMounts := map[string]bool{
+		"/proc":              true,
+		"/dev":               true,
+		"/dev/pts":           true,
+		"/dev/mqueue":        true,
+		"/sys":               true,
+		"/sys/fs/cgroup":     true,
+		"/dev/shm":           true,
+		"/etc/resolv.conf":   true,
+		"/etc/hostname":      true,
+		"/run/secrets":       true,
+		"/run/.containerenv": true,
+	}
+
+	for _, m := range dumpSpec.Mounts {
+		// Following mounts are ignored as they might point to the
+		// wrong location and if ignored the mounts will correctly
+		// be setup to point to the new location.
+		if ignoreMounts[m.Destination] {
+			continue
+		}
+		mount := &types.Mount{
+			ContainerPath: m.Destination,
+			HostPath:      m.Source,
+		}
+
+		for _, createMount := range createMounts {
+			if createMount.ContainerPath == m.Destination {
+				mount.HostPath = createMount.HostPath
+			}
+		}
+
+		for _, opt := range m.Options {
+			switch opt {
+			case "ro":
+				mount.Readonly = true
+			case "rprivate":
+				mount.Propagation = types.MountPropagation_PROPAGATION_PRIVATE
+			case "rshared":
+				mount.Propagation = types.MountPropagation_PROPAGATION_BIDIRECTIONAL
+			case "rslaved":
+				mount.Propagation = types.MountPropagation_PROPAGATION_HOST_TO_CONTAINER
+			}
+		}
+
+		logrus.Debugf("Adding mounts %#v", mount)
+		containerConfig.Mounts = append(containerConfig.Mounts, mount)
+	}
+	sandboxConfig := &types.PodSandboxConfig{
+		Metadata: &types.PodSandboxMetadata{
+			Name:      sb.Metadata().Name,
+			Uid:       sb.Metadata().Uid,
+			Namespace: sb.Metadata().Namespace,
+			Attempt:   sb.Metadata().Attempt,
+		},
+		Linux: &types.LinuxPodSandboxConfig{},
+	}
+
+	if err := ctr.SetConfig(containerConfig, sandboxConfig); err != nil {
+		return "", fmt.Errorf("setting container config: %w", err)
+	}
+
+	if err := ctr.SetNameAndID(ctrID); err != nil {
+		return "", fmt.Errorf("setting container name and ID: %w", err)
+	}
+
+	if _, err = s.ReserveContainerName(ctr.ID(), ctr.Name()); err != nil {
+		return "", fmt.Errorf("kubelet may be retrying requests that are timing out in CRI-O due to system load: %w", err)
+	}
+
+	defer func() {
+		if retErr != nil {
+			log.Infof(ctx, "CreateCtr: releasing container name %s", ctr.Name())
```

Comment:
nit: shouldn't these errors be RestoreCtr

Reply:
Fixed